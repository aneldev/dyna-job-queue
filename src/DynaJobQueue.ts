export interface IDynaJobQueueConfig {
  parallels?: number;   // Default is 1
}

export interface IDynaJobQueueStats {
  jobs: number;
  running: number;
}

interface IQJob {
  priority: number;
  callback: Function;
  internalPriority: number,
}

export class DynaJobQueue {
  private _jobs: IQJob[] = [];
  private _parallels: number = 0;
  private readonly _completeCallbacks: any[] = [];

  constructor(private _config: IDynaJobQueueConfig = {}) {
    this._config = {
      parallels: 1,
      ...this._config
    };
  }

  public addJobPromise<TResolve>(callback: (resolve: (data?: TResolve) => void, reject: (error?: any) => void) => void, priority: number = 1): Promise<TResolve> {
    return new Promise((resolve: (data: any) => void, reject: (error: any) => void) => {
      this.addJobCallback(
        (done: Function) => callback(
          (data: any) => {
            resolve(data);
            done();
          },
          (error: any) => {
            reject(error);
            done();
          }),
        priority);
    });
  }

  public addJobPromised<TResolve>(returnPromise: () => Promise<TResolve>, priority: number = 1): Promise<TResolve> {
    return new Promise((resolve: (resolveData: TResolve) => void, reject: (error: any) => void) => {
      this.addJobCallback(
        (done: () => void) => {
          returnPromise()
            .then((resolveData: TResolve) => {
              resolve(resolveData);
              done();
            })
            .catch((error: any) => {
              reject(error);
              done();
            });
        },
        priority);
    });
  }

  public addJobCallback(callback: (done: Function) => void, priority: number = 1): void {
    this.addJob(priority, callback);
  }

  public jobFactory<TResolve>(func: (...params: any[]) => Promise<TResolve>, priority: number = 1): () => Promise<TResolve> {
    return (...params: any[]) =>
      this.addJobPromised(() => func(...params), priority);
  }

  public get stats(): IDynaJobQueueStats {
    return {
      jobs: this._jobs.length,
      running: this._parallels,
    };
  }

  public get isWorking(): boolean {
    return !!this._jobs.length || !!this._parallels;
  }

  public async allDone(): Promise<void> {
    if (!this.isWorking) return;
    return new Promise(resolve => this._completeCallbacks.push(resolve));
  }

  private addJob(priority: number = 1, callback: (done: Function) => void): void {
    let job: IQJob = {priority, internalPriority: this._createPriorityNumber(priority), callback};
    this._jobs.push(job);
    this._jobs.sort((jobA: IQJob, jobB: IQJob) => jobA.internalPriority - jobB.internalPriority);
    setTimeout(() => this._execute(), 0);
  }

  private _execute(): void {
    if (this._parallels === this._config.parallels) return;
    const jobToExecute: IQJob | undefined= this._jobs.shift();
    if (this._jobs.length === 0) this._internalCounter = 0;

    if (jobToExecute) {
      this._parallels++;
      jobToExecute.callback(() => {
        this._parallels--;
        this._execute();
      });
    }
    else {
      while (this._completeCallbacks.length) this._completeCallbacks.shift()();
    }
  }

  private _internalCounter: number = 0;

  private _createPriorityNumber(priority: number): number {
    return Number(("000000000000000" + priority).substr(-15) + '0' + ("0000000000" + (++this._internalCounter)).substr(-10));
  }

}
