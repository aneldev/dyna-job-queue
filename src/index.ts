export interface IQJob {
  command: string;
  data: any;
  priority: number;
  _internalPriority: number,
}

export class DynaJobQueue {
  private _jobs: IQJob[] = [];
  private _isExecuting: boolean = false;

  public addJob(command: string, data: any, priority: number = 1): IQJob {
    let job: IQJob = {command, data, priority, _internalPriority: this._createPriorityNumber(priority)};
    this._jobs.push(job);
    this._jobs.sort((jobA: IQJob, jobB: IQJob) => jobA._internalPriority - jobB._internalPriority);
    setTimeout(() => this._execute(), 0);
    return job;
  }

  public onJob(job: IQJob, done: () => void): void {
    // to override!
    throw Error('DynaJobQueue: onJob! error, you should override the onJob function where is called when a job is available');
  }

  public get count(): number {
    return this._jobs.length;
  }

  private _execute(): void {
    if (this._isExecuting) return;
    const jobToExecute: IQJob = this._jobs.shift();

    if (jobToExecute) {
      this._isExecuting = true;
      this.onJob(jobToExecute, () => {
        this._isExecuting = false;
        this._execute();
      });
    }
  }

  private _internalCounter: number = 0;

  private _createPriorityNumber(priority: number): number {
    return Number(("000000000000000" + priority).substr(-15) + '0' + (++this._internalCounter));
  }

}
