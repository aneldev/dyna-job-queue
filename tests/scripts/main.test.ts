import "jest";

import {count} from "dyna-count";

import {DynaJobQueue} from '../../src';

// help: https://facebook.github.io/jest/docs/expect.html

describe('Dyna Job Queue - using addJobCallback()', () => {

  let queue = new DynaJobQueue();
  const testCollectedData: string[] = [];


  it('expects the pending jobs should 0', () => {
    expect(queue.stats.jobs).toBe(0);
  });

  it(`should push 10 jobs`, () => {
    count(10).for((index: number) => {
      queue.addJobCallback((done: () => undefined) => {
        testCollectedData.push(`start ${index}`);
        setTimeout(() => {
          testCollectedData.push(`end ${index}`);
          done();
        }, 300);
      });
    });
  });

  it(`expects the pending jobs to be 9 and one is started to working`, () => {
    expect(queue.stats.jobs).toBe(9);
    expect(testCollectedData.length).toBe(1);
  });

  it(`should pick last 10 job in correct order`, (done: () => undefined) => {
    setTimeout(() => {
      const expected: string[] =
        count(10)
          .reduce((acc: string[], index: number) => {
            acc.push(`start ${index}`);
            acc.push(`end ${index}`);
            return acc;
          }, []);
      expect(JSON.stringify(testCollectedData, null, 2))
        .toBe(JSON.stringify(expected, null, 2));
      done();
    }, (10 * 300) + 500);
  });

  it('expects the pending jobs should 0', () => {
    expect(queue.stats.jobs).toBe(0);
  });

});

describe('Dyna Job Queue - using addJobPromise()', () => {
  let queue = new DynaJobQueue();
  const testCollectedData: any[] = [];

  it(`should push 10 jobs`, (done: () => undefined) => {
    for (let i: number = 0; i < 10; i++) {
      queue.addJobPromise((resolve) => {
        setTimeout(() => {
          let data: any = {index: i};
          testCollectedData.push(data);
          expect(queue.isWorking).toBe(true);
          resolve(data);
        }, 100);
      }, 2)
        .then((data: any) => {
          let lastIndexValue: number = testCollectedData[testCollectedData.length - 1].index;
          expect(lastIndexValue).toBe(i);
          expect(data.index).toBe(i);
          if (lastIndexValue == 9) expect(queue.isWorking).toBe(false);
          if (lastIndexValue == 9) done();
        });
    }
  });

  it('expects the pending jobs should 0', () => {
    expect(queue.stats.jobs).toBe(0);
  });

});

describe('Dyna Job Queue - using addJobPromised()', () => {
  let queue = new DynaJobQueue();
  const testCollectedData: any[] = [];

  it(`should push 10 jobs`, (done: () => undefined) => {
    for (let i: number = 0; i < 10; i++) {
      queue.addJobPromised(() => {
        return new Promise((resolve: (date: any) => void) => {
          setTimeout(() => {
            let data: any = {index: i};
            testCollectedData.push(data);
            expect(queue.isWorking).toBe(true);
            resolve(data);
          }, 100);
        });
      }, 2)
        .then((data: any) => {
          let lastIndexValue: number = testCollectedData[testCollectedData.length - 1].index;
          expect(lastIndexValue).toBe(i);
          expect(data.index).toBe(i);
          if (lastIndexValue == 9) expect(queue.isWorking).toBe(false);
          if (lastIndexValue == 9) done();
        });
    }
  });

  it('expects the pending jobs should 0', () => {
    expect(queue.stats.jobs).toBe(0);
  });

});

describe('Dyna Job Queue - using parallels', () => {
  let queue = new DynaJobQueue({parallels: 3});
  let times: { [index: string]: number };

  it('should push 5 jobs', (done: () => undefined) => {
    times = {};
    const now: number = Number(new Date);
    const getNow = (): number => Number(new Date) - now;
    count(5).for((index: number) => {
      queue.addJobPromise((resolve) => {
        times[index] = getNow();
        setTimeout(resolve, 1000);
      });
    });
    setTimeout(done, 2100);
  });

  it('should have the correct times', () => {
    expect(times[0] < 500).toBe(true);
    expect(times[1] < 500).toBe(true);
    expect(times[2] < 500).toBe(true);
    expect(times[3] > 1000).toBe(true);
    expect(times[4] > 1000).toBe(true);
  });

  it('should push 5 jobs (again)', (done: () => undefined) => {
    times = {};
    const now: number = Number(new Date);
    const getNow = (): number => Number(new Date) - now;
    count(5).for((index: number) => {
      queue.addJobPromise((resolve) => {
        times[index] = getNow();
        setTimeout(resolve, 1000);
      });
    });
    setTimeout(done, 1100);
  });

  it('should have the correct times', () => {
    expect(times[0] < 500).toBe(true);
    expect(times[1] < 500).toBe(true);
    expect(times[2] < 500).toBe(true);
    expect(times[3] > 1000).toBe(true);
    expect(times[4] > 1000).toBe(true);
  });
});

describe('Dyna Job Queue - jobFunction', () => {
  let queue = new DynaJobQueue({parallels: 3});

  class NewsFeeder {
    private readonly feeds: number[] = [];

    constructor() {
      this.addFeed = queue.jobFactory(this.addFeed.bind(this));
    }

    public addFeed(feed: number, afterDelay: number): Promise<number> {
      return new Promise((resolve) => {
        setTimeout(() => {
          this.feeds.push(feed);
          resolve(feed);
        }, afterDelay);
      });
    }

    public getFeeds(): number[] {
      return this.feeds;
    }

    public clearFeeds(): void {
      while (this.getFeeds().pop()) {
      }
    }
  }

  const newsFeeder = new NewsFeeder();

  it('adds an item to the feeder', (done: () => undefined) => {
    newsFeeder.addFeed(12, 200)
      .then(() => newsFeeder.getFeeds())
      .then(feeds => {
        expect(feeds.length).toBe(1);
        expect(feeds[0]).toBe(12);
        newsFeeder.clearFeeds();
        done();
      })
  });

  it('adds items with different delay', (done: () => undefined) => {
    Promise.all([
      newsFeeder.addFeed(110, 100),
      newsFeeder.addFeed(105, 200),
      newsFeeder.addFeed(100, 2),
    ])
      .then((executed) => {
        expect(executed.join()).toBe("110,105,100");
        expect(newsFeeder.getFeeds().join()).toBe("100,110,105");
        done();
      });
  });
});

describe('Dyna Job Queue - allDone() at the end', () => {
  const queue = new DynaJobQueue({parallels: 1});
  let text = "";

  let addText = (subText: string) => {
    return new Promise<void>(resolve => {
      setTimeout(() => {
        text += subText;
        resolve();
      }, 200)
    });
  };
  addText = queue.jobFactory(addText);

  it('the allDone() is called on time', async () => {
    addText('1');
    addText('2');
    addText('3');
    await queue.allDone();
    text += 'Done';
    expect(text).toBe('123Done');
  });

});

describe('Dyna Job Queue - allDone() after first job', () => {
  const queue = new DynaJobQueue({parallels: 1});
  let text = "";

  let addText = (subText: string) => {
    return new Promise<void>(resolve => {
      setTimeout(() => {
        text += subText;
        resolve();
      }, 200)
    });
  };
  addText = queue.jobFactory(addText);

  it('the allDone() is called on time', (done) => {
    addText('1');
    queue.allDone()
      .then(() => {
        text += 'Done';
        expect(text).toBe('123Done');
        done();
      });
    addText('2');
    addText('3');
  });

});

describe('Dyna Job Queue - jobFactory massive calls', () => {
  const queue = new DynaJobQueue({parallels: 1});
  let collection: string[] = [];

  let addText = async (text: string) => {
    collection.push(text);
    await new Promise(r => setTimeout(r, 0))
  };

  addText = queue.jobFactory(addText);

  it('the jobs are called quickly on massive calls', async () => {
    const started = Date.now();

    await new Promise(r => setTimeout(r, 100));

    await Promise.all(
      count(20)
        .map(index => addText(`id-${index}`))
    );

    await new Promise(r => setTimeout(r, 100));

    await Promise.all(
      count(20)
        .map(index => addText(`id-${index}`))
    );

    const ended = Date.now();

    const elapsed = ended - started;
    console.log('Elapsed', elapsed);

    expect(elapsed).toBeLessThan(500);
    expect(collection).toMatchSnapshot();
  });

});


describe('Test output of the addJobPromised', () => {
  it('Parallel 1', async () => {
    const queue = new DynaJobQueue({parallels: 1});
    const fileNames = getRowArray(5).map(n => String(n + 1));
    const uploadedFiles: string[] = [];

    await Promise.all(
      fileNames
        .map(fileName  => queue.addJobPromised(() => uploadFile(fileName, uploadedFiles)))
    );

    expect(uploadedFiles).toMatchSnapshot('Uploaded files');
  });

  const uploadFile = async (fileName: string, files: string[]): Promise<string> => {
    console.debug('Uploading file', fileName);
    await new Promise(r => setTimeout(r, 200));
    files.push(fileName + '.txt');
    return 'Uploading file ' + fileName + ' completed';
  };

  const getRowArray = (count: number): number[] =>
    Array(count)
      .fill(null)
      .map((v, index) => {
        v; // 4TS
        return index;
      });
});
