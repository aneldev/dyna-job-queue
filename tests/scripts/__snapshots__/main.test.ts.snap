// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dyna Job Queue - jobFactory massive calls the jobs are called quickly on massive calls 1`] = `
Array [
  "id-0",
  "id-1",
  "id-2",
  "id-3",
  "id-4",
  "id-5",
  "id-6",
  "id-7",
  "id-8",
  "id-9",
  "id-10",
  "id-11",
  "id-12",
  "id-13",
  "id-14",
  "id-15",
  "id-16",
  "id-17",
  "id-18",
  "id-19",
  "id-0",
  "id-1",
  "id-2",
  "id-3",
  "id-4",
  "id-5",
  "id-6",
  "id-7",
  "id-8",
  "id-9",
  "id-10",
  "id-11",
  "id-12",
  "id-13",
  "id-14",
  "id-15",
  "id-16",
  "id-17",
  "id-18",
  "id-19",
]
`;

exports[`Test output of the addJobPromised Parallel 1: Uploaded files 1`] = `
Array [
  "1.txt",
  "2.txt",
  "3.txt",
  "4.txt",
  "5.txt",
]
`;
